### PR TITLE
DYN-5405 npm pack dependencies

### PR DESCRIPTION
**Pull request by filipeotero**
### Purpose

This PR is to avoid installing all the npm dependencies of the Notification Center and Splash Screen.
* Use of npm pack instead of npm install
* Extract the downloaded .tgz file
* Update the assembly path of the files 
* Update the .gitignore for the new files

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@QilongTang 

### FYIs

@RobertGlobant20

### CODE REVIEW

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (QilongTang) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj:20]: @filipeotero Curious if this is depending on any particular zipping software? If I remember correctly, Windows now comes with unzipping tool OOTB?

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (QilongTang) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj:26]: this means?

File: .gitignore
Comment (filipeotero) [inline .gitignore:107]: Removing the old folder hierarchy

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (filipeotero) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj:17]: The first step is to download the .tgz file

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (filipeotero) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj:22]: This gets all the files that contain the .tgz suffix

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (filipeotero) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj:26]: Reverts the order of the found files

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (filipeotero) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj:30]: Gets the last one

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (QilongTang) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:17]: Would you add inline comments, you should be able to use `<!-- comment -->`

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (filipeotero) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:20]: Looks like it's coming with libarchive:
https://learn.microsoft.com/en-us/virtualization/community/team-blog/2017/20171219-tar-and-curl-come-to-windows

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (filipeotero) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:26]: Reverts the order of the files to get the latest version

Comment (QilongTang) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:29]: Would this still work if this project is depending on more than one npm package?

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (QilongTang) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:18]: Should we clean up none latest versioned npm package?

**Review by QilongTang — COMMENTED** [review]
No doubt this would work, but I have some concern. 
1. If we want to specify the npm package version, similar to versioned Nuget package
2. If we want to clean up the legacy npm package.

Just two thoughts I would like team to think over

**Comment by filipeotero** [comment]
> No doubt this would work, but I have some concern.
> 
> 1. If we want to specify the npm package version, similar to versioned Nuget package. If we stick to latest, it seems hard to get the exact package
> 2. If we want to clean up the legacy npm package.
> 
> Just two thoughts I would like team to think over

1 - If we delete it after being extracted, maybe it won't be an issue since we will not maintain the files locally.

**Review by QilongTang — APPROVED** [review]

**Comment by reddyashish** [comment]
The self-service job is running in a loop and the build is failing. Converting this to draft to unblock the jobs.

**Comment by filipeotero** [comment]
We are getting this message from the self-service job:

```
18:27:12    tar -xzf c:\WorkspaceDynamo\src\DynamoCoreWpf\.\dynamods-splash-screen-1.0.5.tgz
18:27:13    'tar' is not recognized as an internal or external command,
18:27:13    operable program or batch file.
18:27:13  c:\WorkspaceDynamo\src\DynamoCoreWpf\DynamoCoreWpf.csproj(38,9): error MSB3073: The command "tar -xzf c:\WorkspaceDynamo\src\DynamoCoreWpf\.\dynamods-splash-screen-1.0.5.tgz" exited with code 9009.
```
Looks like there is no libarchive or something else to extract the tgz file. 
Any ideas on how we can extract the file? 

@QilongTang @reddyashish @RobertGlobant20

**Comment by QilongTang** [comment]
@filipeotero Should we include this dll in the repo, is that enought to leverage it?

**Review by mjkkirschner — CHANGES_REQUESTED** [review]
@filipeotero you cannot include an LGPL license'd component in our repo without consulting legal. I would be hesitant to even reference it, it will only be a continuous source of questions.

**Comment by filipeotero** [comment]
> If this is really the best solution to this problem - then why include the binaries directly instead of referencing nuget packages like: https://www.nuget.org/packages/7-Zip.StandaloneConsole.x64 ?

Here are the things that I tried:
**1 - Extract the file with `tar` and don't insert any binaries.**
Error from the job:
`The command "tar -xzf c:\WorkspaceDynamo\src\DynamoCoreWpf\dynamods-splash-screen-1.0.5.tgz" exited with code 9009.`

**2 - Insert tar.exe from the windows system in the project**
Error from the job:
`The command "../../tar/tar.exe -xzf c:\WorkspaceDynamo\src\DynamoCoreWpf\dynamods-splash-screen-1.0.5.tgz" exited with code -1073741511`

**3 - tar.exe from https://gnuwin32.sourceforge.net/packages/gtar.htm**
Error from the job:
`The command "..\..\extern\tar\tar.exe -xzf c:\WorkspaceDynamo\src\DynamoCoreWpf\dynamods-splash-screen-1.0.6.tgz" exited with code -1073741515.`

**4 - Now from 7zip**

Now I will check this nugget package.

**Comment by mjkkirschner** [comment]
`9009` - is usually something like command not found, how are you pointing to `tar`?
have you tested it locally? Does it work if you use the full path to the tool?

I see a weird combination of / and \ in your commands, maybe those are not exactly what you ran but I would not expect that to work - `\` only on windows.

can you try with `v` for verbose mode, curious if theres any more interested output...

**Comment by filipeotero** [comment]
**how are you pointing to `tar`?** 

On the first attempt, I just called the command directly
Second I pointed to the `extern` folder (To which I added the tar.exe file)

**Have you tested it locally?** 
Yes

**Does it work if you use the full path to the tool?** 
Yes

**From what terminal are you running this?**
Windows command prompt 

 **Have you tried adding tar to your path?**
tar is already included in my PC
check this: https://learn.microsoft.com/en-us/virtualization/community/team-blog/2017/20171219-tar-and-curl-come-to-windows

**can you try with `v` for verbose mode, curious if theres any more interesting output...**
Sure, I will do that.

**Comment by mjkkirschner** [comment]
@filipeotero so - I'm still missing some information.
* does the full build work locally or just your calls to invoke `tar` at the command line? Does this only fail on the build machine? (ie inside docker?)
* IE are you certain that the targets, npm pack steps, and reversal of files are executing in the order you expect?
* I think it would be good to use a message task to log some of the variables you have created, like the last and path etc to see what is getting passed to `tar` - https://learn.microsoft.com/en-us/visualstudio/msbuild/message-task?view=vs-2022 if you don't already see that in the log - you could also try using diagnostic build mode.
*

**Comment by filipeotero** [comment]
**does the full build work locally or just your calls to invoke `tar` at the command line?** 
Both work locally. 

**Does this only fail on the build machine? (ie inside docker?)**
Yes, 

**IE are you certain that the targets, npm pack steps, and reversal of files are executing in the order you expect?**
I guess so because it wouldn't even print the name of the file (dynamods-splash-screen-1.0.6.tgz) if it wasn't executing in the right order. The flow is the following: Download the package -> Find the highest version -> Extract the package -> Delete the package 

**I think it would be good to use a message task to log some of the variables you have created, like the last and path etc to see what is getting passed to `tar` - https://learn.microsoft.com/en-us/visualstudio/msbuild/message-task?view=vs-2022 if you don't already see that in the log - you could also try using diagnostic log mode in visual studio.**
Ok, I will add some messages in the steps. Thanks!

**Comment by mjkkirschner** [comment]
@filipeotero what version of visual studio are you using locally? The builder image in docker is a bit old and based on an old windowsservercore image - perhaps we should update it to something more modern...

**Comment by mjkkirschner** [comment]
@filipeotero @avidit (not sure what investigation you did of the builder image) but
two thoughts:

this post makes it seem like windows doesn't include tar until pretty recently - ... and we're using a build image from 2019.
https://learn.microsoft.com/en-us/virtualization/community/team-blog/2017/20171219-tar-and-curl-come-to-windows

In addition, have you considered using powershell? (which we know works on the builder image)
https://scatteredcode.net/download-and-extract-gzip-tar-with-powershell/

**Comment by mjkkirschner** [comment]
one more thought - if 7zip turns out to be the best solution, another approach is to add it to the builder image - this is nice because we add it for all our jobs that use that image, not just dynamo, in addition it makes it very unlikely we would distribute it - worth checking with legal if that makes things simpler.

**Comment by filipeotero** [comment]
Hi @mjkkirschner, I'm using VS 2019.
What about using the nuget package as you mentioned? I didn't try it yet but if it works we won't need to include binaries, right?

**Comment by mjkkirschner** [comment]
we don't need to include the binaries, but we still need to reference it and still need to go through legal.

**Comment by avidit** [comment]
> @avidit (not sure what investigation you did of the builder image)

**Comment by filipeotero** [comment]
@mjkkirschner I submitted a new commit that deletes the binaries and references the nuget package. Also, I'm still trying to reach the legal team about using 7zip. Thanks!

**Comment by filipeotero** [comment]
@QilongTang Just to let you know, we have to wait for the EngOps to catch up before we merge this. https://jira.autodesk.com/browse/ESC-34251

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (QilongTang) [inline src/DynamoCoreWpf/DynamoCoreWpf.csproj:57]: If there is no need for a specific package name, does it mean that we would only support a single npm package per project? Do we need to specify the folder name during unzipping?

**Review by QilongTang — COMMENTED** [review]
LGTM but with one concern commented

File: src/DynamoCoreWpf/DynamoCoreWpf.csproj
Comment (filipeotero) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:57]: Is not necessary to specify the folder name during unzipping. By default, the folder created is called _package_. We can create folders named properly using the _--directory_ argument from tar.

Comment (QilongTang) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:57]: @filipeotero Understood, my point is that if in the future `DynamoCoreWpf` is depending on multiple npm packages, they will all be unzipped to the `package` folder and the index.html will be overwriting each other. If possible, I would love to have the dir specified to prevent that case from happening, what do you think?

Comment (filipeotero) [reply src/DynamoCoreWpf/DynamoCoreWpf.csproj:57]: Makes sense. I will do that!

**Comment by sm6srw** [comment]
@filipeotero I think we can get this PR ready for merging again. The BRE pipeline have been using VS2022 now for a couple of days.

**Comment by filipeotero** [comment]
@zeusongit Just to let you know, this PR is ready to be merged after the checks are completed. Thanks!

**Pull request merged**
